webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,108 @@
-  return (
-    <ExperimentsContext.Provider value={{ experimentsCache, updateExperimentsCache }}>
-      {children}
-    </ExperimentsContext.Provider>
-  );
-};
+// This is singleton class
+export class ExperimentsService {
+  static #isInternalConstructing = false;
+  static #instance: ExperimentsService;
+  #experimentsCache: WritableObservable<Map<string, ExperimentsCache>> = observable(Map());
+  #experimentMap: WritableObservable<Map<number, ExperimentItem>> = observable(Map());
 
-export const useFetchExperiments = (
-  params: GetExperimentsParams,
-  canceler: AbortController,
-): (() => Promise<void>) => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  private constructor() {
+    if (!ExperimentsService.#isInternalConstructing) {
+      throw new TypeError('ExperimentsService is not constructable');
+    }
   }
-  const { experimentsCache, updateExperimentsCache } = context;
-
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getExperiments(params, { signal: canceler.signal });
-      updateExperimentsCache(experimentsCache.set(encodeParams(params), response));
-    } catch (e) {
-      handleError(e);
+
+  // Get singleton instance
+  public static getInstance(): ExperimentsService {
+    if (!ExperimentsService.#instance) {
+      ExperimentsService.#isInternalConstructing = true;
+      ExperimentsService.#instance = new ExperimentsService();
+      ExperimentsService.#isInternalConstructing = false;
     }
-  }, [canceler, experimentsCache, params, updateExperimentsCache]);
-};
+    return ExperimentsService.#instance;
+  }
 
-export const useExperiments = (params: GetExperimentsParams): Loadable<ExperimentPagination> => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useExperiments outside of Experiment Context');
+  // Get an experiment by experiment id
+  public getExperimentsByIds(experimentIds: number[]): Observable<Loadable<ExperimentItem[]>> {
+    return this.#experimentMap.select((map) => {
+      const expList: ExperimentItem[] = experimentIds
+        .map((id) => map.get(id))
+        .flatMap((exp) => (exp ? [exp] : []));
+      return Loaded(expList);
+    });
   }
-  const loadedVal = context.experimentsCache.get(encodeParams(params));
 
-  return loadedVal ? Loaded(loadedVal) : NotLoaded;
-};
+  public getExperimentsByParams(
+    params: Readonly<GetExperimentsParams>,
+  ): Observable<Loadable<ExperimentPagination>> {
+    return this.#experimentsCache.select((map) => {
+      const cache = map.get(encodeParams(params));
+      if (!cache) {
+        return NotLoaded;
+      }
+      const expMap = this.#experimentMap.get();
+      const experiments: ExperimentItem[] = cache.experimentIds
+        .map((id) => expMap.get(id))
+        .flatMap((exp) => (exp ? [exp] : []));
```

Comment:
This seems to work, but I think you could remove the extra loop and speed it up again by doing this instead:
```
const experiments: ExperimentItem[] = cache.experimentIds
        .flatMap((id) => {
             const exp = expMap.get(id);
             return (exp ? [exp] : [])
         });
```